### PR TITLE
Fix for How to ORDER BY or GROUP BY json data #4102

### PR DESCRIPTION
**Pull request by Shiwanee**
@Amruta-Ranade ,
To fix this issue I have added a new section: Group and order `JSONB` values

1. Please let me know your review comments.
2. Also, I would like if someone can do a Technical review for the example that I have added.
3. It would be helpful to add an example for casting the result to a float value; but I would need some help with that.   
    Let me know if such an example is needed or we are good.
4. I have made the changes only for v2.1, let me know if I need to make them for all other releases.

In this file, apart from fixing this issue, I also observed that:
1.  When you execute the command, SELECT user_profile->>'first_name', user_profile->>'location' FROM users; it 
     does not give the column names, but the output printed in the doc shows the names. I think we need to update the query to 
    match the output shown. Could you please confirm this discrepancy and let me know if I need to make any changes there?
2. Let me know if I should make the heading "Create a Table with a JSONB Column" sentence-case?

Thanks,
Shiwanee

### CODE REVIEW

File: v2.1/jsonb.md [hunk 1]
```diff
@@ -183,6 +183,55 @@ You can also use the `->>` operator to return `JSONB` field values as `STRING` v
 
 For the full list of functions and operators we support, see [Functions and Operators](functions-and-operators.html).
 
+### Group and order `JSONB` values
+
+To organize your `JSONB` field values, use the GROUP BY and ORDER BY clauses with the `->>` operator. For example, organise the `first_name` values from the table you created in the [first example](#create-a-table-with-a-jsonb-column):
```
Comment (ericharmeling) [inline v2.1/jsonb.md:188]: > GROUP BY and ORDER BY

`GROUP BY` and `ORDER BY`

File: v2.1/jsonb.md [hunk 2]
```diff
@@ -183,6 +183,55 @@ You can also use the `->>` operator to return `JSONB` field values as `STRING` v
...
++----------------------------+--------------------------+
+| first_name                 | total                    |
++----------------------------+--------------------------+
+| Erine                      | 1                        |
+| Parvati                    | 1                        |
+| Lola                       | 2                        |
++----------------------------+--------------------------+
+~~~
+
+The `->>` operator returns a `STRING` value and uses string comparison rules to order the data. If you want numeric ordering, cast the resulting data to a float.
```
Comment (ericharmeling) [inline v2.1/jsonb.md:231]: > `STRING` value

[`STRING`](string.html)

Comment (ericharmeling) [inline v2.1/jsonb.md:231]: > float

[`FLOAT`](float.html)

**Review by ericharmeling — COMMENTED** [review]
@Shiwanee Thanks so much for your contribution! Please see my inline review comments.

Another request before I can approve the changes: 
Please make documentation updates to v20.1 and v20.2 docsets, using output from the correct installation of CockroachDB (v20.1 for the v20.1 docs and v20.2 for the v20.2 docs). We don't support or maintain v2.1 anymore.

**Comment by Shiwanee** [comment]
@ericharmeling Thanks so much for reviewing the PR. Currently, I have made changes for v20.1; I'll check the output on v20.2 and push another commit.
Also, I have reverted changes for v2.1.
Thanks,
Shiwanee

**Comment by Shiwanee** [comment]
Hi @ericharmeling , I could reproduce the functionality on v20.2. I have made the required changes to the documentation.

I noticed that the operator "->" also works without any error message as opposed to the problem statement (https://forum.cockroachlabs.com/t/order-by-and-group-by-clauses-on-jsonb-data/2211).

Please let me know if this the expected behavior in v20.2 or else I'll make the required changes in the v20.2 documentation.

I have attached the "->" functionality screenshot with this message.

Thanks,
Shiwanee

jsonb_order_by_v202

**Review by ericharmeling — COMMENTED** [review]
Thank you for iterating on this.

The output for the examples still doesn't have the correct formatting though. For example, this is how the first `SELECT` statement should look:

{% include copy-clipboard.html %}
~~~sql
> SELECT user_profile->>'first_name' AS first_name, user_profile->>'location' AS location FROM users;
~~~

~~~
  first_name | location
-------------+-----------
  Lola       | Seoul
  Ernie      | Brooklyn
  Lola       | NYC
  Parvati    | London
(4 rows)
~~~

**Comment by ericharmeling** [comment]
> I noticed that the operator "->" also works without any error message as opposed to the problem statement (https://forum.cockroachlabs.com/t/order-by-and-group-by-clauses-on-jsonb-data/2211).
> 
> Please let me know if this the expected behavior in v20.2 or else I'll make the required changes in the v20.2 documentation.

Good question! This is expected functionality, and actually produces different output from `->>`:

~~~
root@127.0.0.1:58005/defaultdb> SELECT user_profile->>'first_name' AS first_name, user_profile->>'location' AS location FROM users;
~~~
~~~
  first_name | location
-------------+-----------
  Lola       | Seoul
  Ernie      | Brooklyn
  Lola       | NYC
  Parvati    | London
(4 rows)
~~~

~~~
root@127.0.0.1:58005/defaultdb> SELECT user_profile->'first_name' AS first_name, user_profile->'location' AS location FROM users;
~~~
~~~
  first_name |  location
-------------+-------------
  "Lola"     | "Seoul"
  "Ernie"    | "Brooklyn"
  "Lola"     | "NYC"
  "Parvati"  | "London"
(4 rows)
~~~

As you can see, there are double quotes around the values indicate that they are in fact fields of a JSONB value. For an example of this, see https://www.cockroachlabs.com/docs/stable/jsonb.html#retrieve-specific-fields-from-a-jsonb-value.

**Comment by Shiwanee** [comment]
> Thank you for iterating on this.
> 
> The output for the examples still doesn't have the correct formatting though. For example, this is how the first `SELECT` statement should look:
> 
> {% include copy-clipboard.html %}
> 
> ```sql
> > SELECT user_profile->>'first_name' AS first_name, user_profile->>'location' AS location FROM users;
> ```
> 
> ```
>   first_name | location
> -------------+-----------
>   Lola       | Seoul
>   Ernie      | Brooklyn
>   Lola       | NYC
>   Parvati    | London
> (4 rows)
> ```


Thanks so much for your time @ericharmeling in reviewing these files.
I was not sure what you mean by "correct formatting". The only difference I can see is in the way first_name is ordered. I had installed v20.1 again and got that output. 
Also, as I had mentioned in my first comment (https://github.com/cockroachdb/docs/pull/8679#issue-504409880) I do not see column names when I execute some queries. Is there a chance that something is wrong with my installation or setup?

**Comment by Shiwanee** [comment]
> > I noticed that the operator "->" also works without any error message as opposed to the problem statement (https://forum.cockroachlabs.com/t/order-by-and-group-by-clauses-on-jsonb-data/2211).
> > Please let me know if this the expected behavior in v20.2 or else I'll make the required changes in the v20.2 documentation.
> 
> Good question! This is expected functionality, and actually produces different output from `->>`:
> 
> ```
> root@127.0.0.1:58005/defaultdb> SELECT user_profile->>'first_name' AS first_name, user_profile->>'location' AS location FROM users;
> ```
> 
> ```
>   first_name | location
> -------------+-----------
>   Lola       | Seoul
>   Ernie      | Brooklyn
>   Lola       | NYC
>   Parvati    | London
> (4 rows)
> ```
> 
> ```
> root@127.0.0.1:58005/defaultdb> SELECT user_profile->'first_name' AS first_name, user_profile->'location' AS location FROM users;
> ```
> 
> ```
>   first_name |  location
> -------------+-------------
>   "Lola"     | "Seoul"
>   "Ernie"    | "Brooklyn"
>   "Lola"     | "NYC"
>   "Parvati"  | "London"
> (4 rows)
> ```
> 
> As you can see, there are double quotes around the values indicate that they are in fact fields of a JSONB value. For an example of this, see https://www.cockroachlabs.com/docs/stable/jsonb.html#retrieve-specific-fields-from-a-jsonb-value.

@ericharmeling So should I document both the operators for v20.2, or should I just stick to ->> for this particular issue?

**Comment by ericharmeling** [comment]
@Shiwanee 

> @ericharmeling So should I document both the operators for v20.2, or should I just stick to ->> for this particular issue?

I'd just stick with `->>` for this issue! Thanks. :)

**Comment by ericharmeling** [comment]
> Thanks so much for your time @ericharmeling in reviewing these files.

No problem! :)

> I was not sure what you mean by "correct formatting". The only difference I can see is in the way first_name is ordered. I had installed v20.1 again and got that output.
> Also, as I had mentioned in my first comment ([#8679 (comment)](https://github.com/cockroachdb/docs/pull/8679#issue-504409880)) I do not see column names when I execute some queries. Is there a chance that something is wrong with my installation or setup?

The formatting of statement output in the SQL shell changed in 19.2.

So for this statement:

{% include copy-clipboard.html %}
~~~sql
> SELECT user_profile->>'first_name' AS first_name, user_profile->>'location' AS location FROM users;
~~~

This is the old formatting (which you currently have in this PR):

~~~
+----------------------------+--------------------------+
| first_name                 |location                  |
+----------------------------+--------------------------+
| Ernie                      | Brooklyn                 |
| Lola                       | NYC                      |
| Parvati                    | London                   |
| Lola                       | Seoul                    |
+----------------------------+--------------------------+
~~~

And this is the newer format:

~~~
  first_name | location
-------------+-----------
  Lola       | Seoul
  Ernie      | Brooklyn
  Lola       | NYC
  Parvati    | London
(4 rows)
~~~

The borders and sizing are pretty different. We haven't made time to update all of our examples to show the latest formatting (as you can see on this page, and several others), but we want to make sure any new PRs have the latest output.

In the screenshot that you posted (https://github.com/cockroachdb/docs/pull/8679#issuecomment-713838205), it looks like you were getting the correct output format (which implies that you have the correct version installed on your machine), but just not copying and pasting it to your changes in the PR.

**Comment by Shiwanee** [comment]
> > Thanks so much for your time @ericharmeling in reviewing these files.
> 
> No problem! :)
> 
> > I was not sure what you mean by "correct formatting". The only difference I can see is in the way first_name is ordered. I had installed v20.1 again and got that output.
> > Also, as I had mentioned in my first comment ([#8679 (comment)](https://github.com/cockroachdb/docs/pull/8679#issue-504409880)) I do not see column names when I execute some queries. Is there a chance that something is wrong with my installation or setup?
> 
> The formatting of statement output in the SQL shell changed in 19.2.
> 
> So for this statement:
> 
> {% include copy-clipboard.html %}
> 
> ```sql
> > SELECT user_profile->>'first_name' AS first_name, user_profile->>'location' AS location FROM users;
> ```
> 
> This is the old formatting (which you currently have in this PR):
> 
> ```
> +----------------------------+--------------------------+
> | first_name                 |location                  |
> +----------------------------+--------------------------+
> | Ernie                      | Brooklyn                 |
> | Lola                       | NYC                      |
> | Parvati                    | London                   |
> | Lola                       | Seoul                    |
> +----------------------------+--------------------------+
> ```
> 
> And this is the newer format:
> 
> ```
>   first_name | location
> -------------+-----------
>   Lola       | Seoul
>   Ernie      | Brooklyn
>   Lola       | NYC
>   Parvati    | London
> (4 rows)
> ```
> 
> The borders and sizing are pretty different. We haven't made time to update all of our examples to show the latest formatting (as you can see on this page, and several others), but we want to make sure any new PRs have the latest output.
> 
> In the screenshot that you posted ([#8679 (comment)](https://github.com/cockroachdb/docs/pull/8679#issuecomment-713838205)), it looks like you were getting the correct output format (which implies that you have the correct version installed on your machine), but just not copying and pasting it to your changes in the PR.

@ericharmeling Aaha! Thanks for the detailed reply. My bad, I was more focused on the data formatting than the table formatting, and missed your point.

I used the formatting of other tables in the file for consistency.  While I am here, I would be more than happy to fix table format for jsonb.md for both v20.1 and v20.2.. Let me know.

Thanks,
Shiwanee

**Comment by Shiwanee** [comment]
@ericharmeling I have incorporated review comments; please have a look. :-) 

Thanks,
Shiwanee

**Comment by Shiwanee** [comment]
@ericharmeling , do you have any inputs on this? :-)

Thanks,
Shiwanee

**Comment by ericharmeling** [comment]
@Shiwanee 

Looks good! Thanks for going back over this!

**Review by ericharmeling — APPROVED** [review]

**Comment by Shiwanee** [comment]
@ericharmeling Thanks for all the help on this issue. :-)

**Comment by ericharmeling** [comment]
@Shiwanee Of course! Thank you for contributing!

**Pull request merged**